### PR TITLE
GitHub Action for Linting Hyperlink Redirection

### LINKED ISSUE #70: Make Hyperlinks Open a New Tab
### Purpose
Change hyperlinks on pages to open a new tab rather than navigate to a new page away from the docs website. Only do this for external sites.

### Changes
1. If the hyperlink is to an external site, then add `{target=_blank}` to the end of the link. For example:
```
[Python Tutorial](https://docs.python.org/3/tutorial/){target=_blank}
```

2. If the hyperlink is to another page on the docs site, then **leave it be**.

### PR DESCRIPTION
**Pull request by DFriend01**
Resolves #70

### What I Did
I created a GitHub Action that asserts that links to external sites redirect to a new tab using the regular expression defined in #70. Refer to [`./.github/actions/lint`](https://github.com/UBCSailbot/docs/tree/user/dfriend01/70-hyperlinks-action/.github/actions/lint) for the main logic behind the linter. Also check out the [latest run](https://github.com/UBCSailbot/docs/actions/runs/3353829300/jobs/5556927129) where the linter fails because of links that got matched by the regular expression.

### What to do once approved
1. Change all failing links so that linting passes
2. Remove line 7 in [`./.github/workflows/lint.yml`](https://github.com/UBCSailbot/docs/blob/user/dfriend01/70-hyperlinks-action/.github/workflows/lint.yml)

### CODE REVIEW

**Review by patrick-5546 — COMMENTED** [review]
Can you explore whether it would easy to add annotations? Annotations would allow you to see the errors at the lines they were raised when you click on a commit, [here is an example of markdownlint annotation in our repo](https://github.com/UBCSailbot/docs/pull/63/commits/ecfab06acc7473e17ac0c4ff32aa8bb631130765). Don't worry about it if it would take a lot of time.

### Relevant Resources
- https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#using-workflow-commands-to-access-toolkit-functions
- https://tgrall.github.io/blog/2021/11/07/how-to-write-a-github-action-annotation-api

**Comment by DFriend01** [comment]
@patrick-5546 The annotations work! Here is the [latest action run](https://github.com/UBCSailbot/docs/actions/runs/3353936195/jobs/5557122301) and [its annotations](https://github.com/UBCSailbot/docs/pull/86/commits/8e025a5efc751a8e8871cecc1dc1e39ce7d6e11d)

File: .github/actions/lint/lint_links.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    root = './docs/'
+    markdown_files = get_markdown_files(root)
+    bad_link_matcher = re.compile(r"(?<!!)\[.*\]\(\s*https?:\/\/[^\(\)]+\)(?!\{\s*:?\s*target\s*=\s*(?:\s*_blank\s*|\s*\"\s*_blank\s*\"\s*)\})")
+    passed = lint_markdown_files(markdown_files, bad_link_matcher)
+    return passed
+
+if __name__ == '__main__':
+    passed = main()
+    if not passed:
+        sys.exit(1)
```
Comment (patrick-5546) [inline .github/actions/lint/lint_links.py]: I'm not sure if this is how it's supposed to be but this is how I organize my scripts
- helper functions are somewhere below the function that uses it
    - main() is the first function defined
- call main() and main() only at the bottom
    ```python
    if __name__ == '__main__':
        main()
    ```

**Review by patrick-5546 — COMMENTED** [review]
Should we enforce any standard on internal links? Should we error if they open in a new tab, or should this be left up to the discretion of the author?

File: .github/workflows/lint.yml [hunk 2]
```diff
@@ -4,6 +4,7 @@ on:
   push:
     branches:
       - main
+      - user/dfriend01/70-hyperlinks-action
```
Comment (patrick-5546) [inline .github/workflows/lint.yml]: This is duplicate with the pull request trigger. The reason I did it this way is that before creating a PR you can push your branch without running CI. Once you are ready to run CI you can create a PR

Comment (DFriend01) [reply .github/workflows/lint.yml]: Line 7 is now removed. Thank you.

**Comment by DFriend01** [comment]
@patrick-5546 I suggest that we tackle the problem that you mentioned more generally by adding the choice for a config file. That way, users can specify to ignore specific patterns. What do you think?

Take a look at [this resource](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action). We can specify an input, like a path to a config file, and then read it in the python script.

**Comment by patrick-5546** [comment]
Can you give an example of what this would look like for the case I outlined?

**Comment by DFriend01** [comment]
We would add an ignore pattern for links that match the following regular expression:

```regex
https://ubcsailbot.github.io/docs/.*
```

This would be one ignore pattern that we would add to ignore links back to our site.

**Comment by patrick-5546** [comment]
I think you misunderstood the case I was trying to outline. Let me explain it in more detail.

By internal links, I mean when you link to another file in the repository using its relative path, for example:
```
[homepage](../index.md)
```

With that definition out of the way, here are some error cases that I propose we should add:
- External link to site used instead of internal link; for example if we are in `docs/`:
    ```
    [homepage](https://ubcsailbot.github.io/docs/index.md) -> [homepage](./index.md)
    ``` 
- Internal link opens in new tab instead of in the same tab
    ```
   [homepage](./index.md){target=_blank} -> [homepage](./index.md)
    ``` 
    - This could be configurable
- Ignore pattern ike you suggested

Configurability Options
1. In workflow (easiest if there are only a few options)
    https://github.com/UBCSailbot/docs/blob/fe56aa3b940bddfb6f9d36ae789664494e374e4d/.github/workflows/lint.yml#L20-L21
2. Configuration file

When deciding whether to implement these additional error cases, please consider whether it is worth the effort.

**Comment by DFriend01** [comment]
Let's make the pattern configuration into a separate issue because this one is already getting very large.

File: .github/actions/lint/lint_links.py (hunk 1 above)
Comment (DFriend01) [reply .github/actions/lint/lint_links.py]: Done

**Review by patrick-5546 — APPROVED** [review]
Great work on this PR. I created a task list to track improvements to this linter and updated the issue you created.
- #94 
- #88 

On thing I noticed is that there was some extra whitespace in your python script, flake8 would catch that, or you can make whitespace at the end of lines visible in VS Code.

**Pull request merged**